### PR TITLE
asdf: Don't symlink lib and bin files

### PR DESCRIPTION
**Pull request by ericbn**
because asdf will only partially work if the `ASDF_DIR` is resolved to `/usr/local/`. E.g. just running `asdf` will print

    cat: /usr/local/VERSION: No such file or directory
    version:

    cat: /usr/local/help.txt: No such file or directory

    PLUGIN direnv
      asdf direnv
      asdf direnv hook asdf

    "Late but latest"
    -- Rajinikanth

instead of the version and the complete help message.

It should be installed as the instructions in https://asdf-vm.com/#/core-manage-asdf?id=add-to-your-shell as the `ASDF_DIR` is resolved in https://github.com/asdf-vm/asdf/blob/ccf76dd7e60de8a3de03cd3574f570dc1f9dbc55/lib/utils.bash#L24-L29 after utils.bash is sourced in https://github.com/asdf-vm/asdf/blob/ccf76dd7e60de8a3de03cd3574f570dc1f9dbc55/bin/asdf#L4

So to resolve the `ASDF_DIR` correctly as `/usr/local/opt/asdf`, the `/usr/local/opt/asdf/bin/asdf` script must be directly executed.

If lib and bin files were symlinked instead, then `/usr/local/bin/asdf` would be executed, resolving the `ASDF_DIR` incorrectly to `/usr/local`.

- [X] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/HEAD/CONTRIBUTING.md)?
- [X] Have you ensured that your commits follow the [commit style guide](https://docs.brew.sh/Formula-Cookbook#commit)?
- [X] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [X] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [X] Is your test running fine `brew test <formula>`, where `<formula>` is the name of the formula you're submitting?
- [X] Does your build pass `brew audit --strict <formula>` (after doing `brew install --build-from-source <formula>`)? If this is a new formula, does it pass `brew audit --new <formula>`?

-----

### CODE REVIEW

**Comment by ericbn** [comment]
I feel like making the formula keg_only is not the right approach here, but not sure.

How can we make the formula create only the `/usr/local/opt/asdf` and the shell completion symlinks, but not the symlinks in `/usr/local/bin/` and `/usr/local/lib/`?

**Comment by carlocab** [comment]
See #73173, since I think that's related to this.

CC @jthegedus -- I hope you don't mind that I'm tagging you here for your feedback. You'll probably have to field the issues from users asking why `asdf` isn't in their `PATH` after doing `brew install asdf` anyway if/when this is merged.

**Comment by ericbn** [comment]
@carlocab, this is indeed related to #73173. Thanks for pointing that out. The solution there is not correct though -- and mine is definitely incomplete, as it's not going to symlink the shell completion scripts.

**Comment by jthegedus** [comment]
@carlocab Thanks, we would like the heads up when this changes in a breaking way. Though none of the `asdf` core team are Homebrew users, so if the changes here and fixes in our docs were contributed by the same person it would benefit all.

>@carlocab, this is indeed related to #73173. Thanks for pointing that out. The solution there is not correct though -- and mine is definitely incomplete, as it's not going to symlink the shell completion scripts.

@ericbn These types of conversations should be hashed out by other `asdf` via Homebrew users. There was so much information dumped in #73173 and associated PRs that I was unable to follow with my limited Homebrew knowledge what the implications were.

My suggestion would be, that once `asdf`/Homebrew users agree on a change/path forward here, that it be summarised and then myself or @stratus3d looped in to understand the implications on existing user base (which ideally is minimal as Homebrew lists ~90k installs in the last 365 days and that's a lot of systems to break).

**Comment by ericbn** [comment]
The homebrew installation of asdf is partially broken, because it adds an asdf command to the PATH that does not work completely, as it's not located in the right directory. That directory gets resolved to ASDF_DIR, but it's not the right ASDF_DIR.

I still think the fix in homebrew-core is that it should not add this command to the PATH. The workaround is to make sure $ASDF_DIR gets added to the beginning of the PATH, so it clobbers what is added by the homebrew installation. But the installation interferes with what is recommended in https://github.com/asdf-community/asdf-direnv#pro-tips, where $ASDF_DIR/bin gets added to the end of the PATH instead. Also, having a broken command in the PATH is not a good to have anyway.

**Comment by carlocab** [comment]
What about if we install `asdf` into `libexec`, but create a shim in `bin` called `asdf` that just executes the `asdf` that lives in `libexec`? (I feel like this is an alternative that's been discussed before, but it's been a while and I've probably reviewed a few hundred PRs since then...)

**Comment by ericbn** [comment]
@carlocab, your suggestion sounds perfect to me. Let me work on it and update the PR.

**Comment by ericbn** [comment]
@carlocab, PR updated.

@jthegedus, can we have
```
source "$(dirname "$(dirname "${BASH_SOURCE:-$0}")")/lib/utils.bash"
```
instead of
```
source "$(dirname "$(dirname "$0")")/lib/utils.bash"
```
in `bin/asdf`, so it also works when being sourced instead of executed? Otherwise, we're having to do two executions (user executes /usr/local/Cellar/asdf/0.8.1/bin/asdf, which executes /usr/local/Cellar/asdf/0.8.1/libexec/bin/asdf) which adds the overhead of having a second bash process. (Otherwise we could have: user executes /usr/local/Cellar/asdf/0.8.1/bin/asdf, which sources /usr/local/Cellar/asdf/0.8.1/libexec/bin/asdf).

I also noted you banned `realpath` and `readlink` from being used in asdf, so having `bin/asdf` resolve the symlink to the actual file does not seem to be an option. (In this case we would have /usr/local/Cellar/asdf/0.8.1/bin/asdf as a symlink to /usr/local/Cellar/asdf/0.8.1/libexec/bin/asdf)

File: Formula/asdf.rb
Comment (carlocab) [inline Formula/asdf.rb:34]:
```suggestion
    touch libexec/"asdf_updates_disabled"
    bin.write_exec_script libexec/"bin/asdf"
```

Comment (ericbn) [reply Formula/asdf.rb]: Much better now! Thanks a lot @carlocab!

**Comment by carlocab** [comment]
LGTM, but I think there was discussion of whether existing users need a migration path to the new setup in the last PR.

I _think_ this shouldn't lead to any visible changes to most users, but I don't use `asdf` so it's hard to say.

**Comment by ericbn** [comment]
For existing users that are also sourcing the `asdf.sh` initialization file that is mentioned in https://asdf-vm.com/guide/getting-started.html#_3-add-to-your-shell, the instructions will change from

> Add `. $(brew --prefix asdf)/asdf.sh` to your ~/.bashrc or ~/.zshrc file.

to

> Add `. $(brew --prefix asdf)/libexec/asdf.sh` to your ~/.bashrc or ~/.zshrc file.

This step is important for most users, because it enables the support for the asdf shell command, and most importantly, it adds the asdf shims directory to the PATH.

@carlocab, maybe we can add an `/asdf.sh` script that sources `/libexec/asdf.sh` so existing users won't even feel the changes in this PR. WDYT?

**Comment by carlocab** [comment]
Makes sense to me. Do these need to be scripts that source the ones that live in `libexec`, or would symlinks work here?

I suppose symlinks might suffer from the same problem this PR is trying to solve.

**Comment by ericbn** [comment]
Correct, symlinks will make the script fail to resolve the `$0` variable, which is what we're fixing here. Is there a `write_source_script` function like the cool `write_exec_script` one, or do I have to do this one manually?

**Comment by carlocab** [comment]
This one you'll have to write manually.

**Comment by cho-m** [comment]
As a note, with changes, the current PR solution here is similar to #73173.

There was a lot happening there, but the main solution was just:
```rb
    libexec.install Dir["*"]
    touch libexec/"asdf_updates_disabled"
    bin.write_exec_script (opt_libexec/"bin/asdf")
    (prefix/"asdf.sh").write ". #{opt_libexec}/asdf.sh\n"
    (prefix/"asdf.fish").write "source #{opt_libexec}/asdf.fish\n"
```

At minimum, this PR should also should bump the revision so users get the update.

Also, I think this PR can similarly remove `conflicts_with "homeshick"`, which was due to arbitrary ASDF directories/files being copied into main Homebrew prefix.

---

The other changes from the previous PR were:
1. Adding a "setup" caveat using Homebrew paths and aligned to official upstream documentation. This was a point that I had previously recommended going with simple caveat rather than what is in the other PR (i.e. similar to `chruby` and `autojump`)
2. Removing the dependencies. This is a bit trickier whether we should or should not do this. ASDF by itself has pretty much no dependencies; however, many ASDF plugins compile languages from source. So, ASDF plugins (Python/Ruby for example) will usually not work out-of-the-box without packages like autotool/openssl/readline

**Comment by ericbn** [comment]
@cho-m, cool to see two people independently arrived at the same solutions :- )

Added a commit with what was discussed above, except for removing the dependencies.

**Comment by carlocab** [comment]
Cleaned up your commits a bit by reformatting them according to Homebrew/core style.

I'll leave this open for a bit for more feedback from `asdf` users. We should be able to merge this soon.

**Comment by jthegedus** [comment]
Glad everyone is agreeing on a solution :)

As I understand from reading this thread, the main change the core should make is a documentation update to tell new users to

>Add `. $(brew --prefix asdf)/libexec/asdf.sh` to your `~/.bashrc` or `~/.z

Additionally, this update won't affect existing users as it includes a script to rewrite the old entrypoint to the new entrypoint at `libexec/asdf.sh`?

Is my understanding correct?

**Comment by carlocab** [comment]
@jthegedus that is correct.

**Comment by jthegedus** [comment]
Great, I will work on a documentation update and merge that once this is merged. @carlocab hopefully some of the other people you tagged in the asdf Homebrew issue come back to review soon.

**Comment by ericbn** [comment]
@jthegedus, we've made the change so it's non-breaking for existing users. So instructions remain the same, using the `. $(brew --prefix asdf)/asdf.sh` path.

One more couple of files that users might want to source directly are `lib/asdf.{sh, fish}`, as part of the advanced installation of the direnv plugin, as documented here: https://github.com/asdf-community/asdf-direnv/blob/master/README.md#pro-tips

@carlocab, can I also add

    (prefix/"lib/asdf.sh").write ". #{libexec}/lib/asdf.sh\n"
    (prefix/"lib/asdf.fish").write "source #{libexec}/lib/asdf.fish\n"

to this PR?

These are all workarounds to avoid symlinks automatically created by Homebrew when we do `prefix.install Dir["*"]` instead of `libexec.install Dir["*"]`...

**Comment by carlocab** [comment]
That's fine, but make it
```
(lib/"asdf.sh").write
```
etc. 

Squash the change with the first commit when you're done.

**Comment by jthegedus** [comment]
>@jthegedus, we've made the change so it's non-breaking for existing users. So instructions remain the same, using the . $(brew --prefix asdf)/asdf.sh path.

Would we not want to update the instructions to source the `$(brew --prefix asdf)/libexec/asdf.sh` file directly so the rewrite file can be removed in future? Or is that fine to exist ad infinitum?

>One more couple of files that users might want to source directly are lib/asdf.{sh, fish}, as part of the advanced installation of the direnv plugin, as documented here: asdf-community/asdf-direnv@master/README.md#pro-tips

Those instructions won't be included in the core docs site as they are for a specific plugin with it's system deps. If those updates are required for `asdf-direnv` then a PR to their docs would be good. Unless I have misunderstood what you mean?

**Comment by carlocab** [comment]
> Would we not want to update the instructions to source the `$(brew --prefix asdf)/libexec/asdf.sh` file directly so the rewrite file can be removed in future? Or is that fine to exist ad infinitum?

I think we want to update the instructions. The `asdf.sh` that lives outside `libexec` can be there for a (long) while, but it's a bit silly to tell users to source a script that only sources another one.

**Comment by ericbn** [comment]
I think the choices are:

1. Either we make this a breaking change, but simpler (no intermediate "exec" and "source" files), telling users to use the files in libexec directly, which will also require changes in the respective documentations.
2. Or we make this non-breaking, as we're trying to do here (including my suggestion to also add more intermediate "source" files for `lib/asdf.{sh, fish}`), but keep it complex.

I personally would pick (1), if my vote would count. And if, @jthegedus, you're saying that updating the docs is fine, I'm understanding you're fine with (1) too. @carlocab, is this towards what you're preferring too?

**Comment by carlocab** [comment]
Yes. We don't want users to have things break on them on their next `brew upgrade`, but we also don't want these placeholder scripts in the prefix to stick around forever (and most users would probably prefer to source the file directly than the placeholder).

**Comment by jthegedus** [comment]
My vote is also for 1+2 solution. I am not sure how long, but at some point in the future we can just remove these extra scripts. It will definitely break someone's machine, but if we give it a year or so it will be a small subset.

**Comment by ericbn** [comment]
@carlocab, just added the fixup to the existing commit.

**Comment by carlocab** [comment]
Forgot to remind you to add a `revision` so we can ship this fix to existing users. Also added a date to the comment so we know when to remove it.

File: Formula/asdf.rb
Comment (Stratus3D) [inline Formula/asdf.rb:7]: Wouldn't this really be revision 2?

**Review by Stratus3D — DISMISSED** [review]

File: Formula/asdf.rb
Comment (ericbn) [inline Formula/asdf.rb:29]: Shouldn't we consider this a placeholder file too (to be included below)?

File: Formula/asdf.rb
Comment (carlocab) [reply Formula/asdf.rb:7]: An omitted `revision` is implicitly `revision 0`.

File: Formula/asdf.rb
Comment (carlocab) [reply Formula/asdf.rb:29]: I think we want users to always be able to call `asdf` even if they don't source the setup scripts (even if this means limited functionality), so we'll still want to keep this even when we remove the rest.

Am I mistaken here? If so, we'll probably want to mention in the `caveats` that the additional setup is now mandatory for anything to work at all.

Comment (ericbn) [reply Formula/asdf.rb:29]: @carlocab, actually the additional setup was always mandatory, and still is. Having just asdf on the path, and being able to run it and see the help message output, might give users the false impression that they already have it set up, but the setup script is required. For example, the script adds the shims directory to the path, which is absolutely necessary for asdf to work.

Comment (carlocab) [reply Formula/asdf.rb:29]: Still, some things work without them. @jthegedus, how common is it for users to find themselves misled by having `asdf` in their `PATH` without having done the proper setup?

Comment (carlocab) [reply Formula/asdf.rb:29]: I can believe that the lion's share of those users installed asdf using Homebrew -- that doesn't tell me how often that happens though.

On the other hand, I feel like we might be trading one issue for another here. That is, we might get fewer users who never sourced the set up scripts and are confused by asdf not working, but we also might get more users who did `brew install asdf` and are confused about why calling `asdf` returns `command not found`.

Not sure which one's better -- might work out to be effectively the same either way.

Comment (ericbn) [reply Formula/asdf.rb:29]: Not sure I get which are the two issues we're trading here. The "users who never sourced the set up scripts" never had and won't have asdf working as expected. If they try to call asdf and get a "command not found" error, it's just a different symptom that they didn't source the set up scripts yet, and don't have asdf working as expected either due to that. We're just letting it 
"fail fast" now, which I would argue is better!  :- )  Also, we're adding the caveats, in the hope users will read that -- if they didn't read any asdf docs/repo etc., because they "just run brew install asdf without any research", at least they have the caveats, which is typically what Homebrew users look at.

Comment (jthegedus) [reply Formula/asdf.rb:29]: I do not have concrete numbers. I agree with @ericbn here and is why, in many a prior PR, I was trying to get caveats instituted to alert users.

Comment (carlocab) [reply Formula/asdf.rb:29]: Ok, fair enough, if there truly cannot be a user with a use-case for doing `brew install asdf` then calling `asdf` for some reason without sourcing the setup scripts, we can move this line back down.

**Review by carlocab — APPROVED** [review]

**Review by Stratus3D — APPROVED** [review]

**Pull request closed without merging**
